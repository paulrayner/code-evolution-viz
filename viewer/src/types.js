/**
 * Shared types (mirrors processor types)
 */
export {};
