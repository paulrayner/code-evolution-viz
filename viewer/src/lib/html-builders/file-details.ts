/**
 * Pure HTML builder for file details panel
 * Extracted from main.ts showFileDetails function
 */

import { FileNode } from '../../types';
import { Cluster, CouplingEdge } from '../../coupling-types';

export interface FileDetailsData {
  file: FileNode;
  githubFileUrl: string | null;
  commitInfo: {
    commitHashStr: string; // Short hash (7 chars)
    message: string;
    siblings: FileNode[]; // Other files in same commit (excluding current file)
  } | null;
  clusterInfo: {
    cluster: Cluster;
    topEdges: CouplingEdge[]; // Already sorted and limited to top 5
  } | null;
}

/**
 * Build HTML for file details panel
 * Pure function - no side effects, no DOM access, no global state
 */
export function buildFileDetailsHTML(data: FileDetailsData): string {
  const { file, githubFileUrl, commitInfo, clusterInfo } = data;

  // Format last modified date
  const lastModifiedStr = file.lastModified
    ? new Date(file.lastModified).toLocaleDateString('en-US', {
        year: 'numeric',
        month: 'short',
        day: 'numeric'
      })
    : 'Unknown';

  const authorStr = file.lastAuthor || 'Unknown';
  const commitHashStr = file.lastCommitHash ? file.lastCommitHash.substring(0, 7) : 'Unknown';

  // Format file age
  const fileAgeStr = file.firstCommitDate
    ? (() => {
        const now = Date.now();
        const fileDate = new Date(file.firstCommitDate).getTime();
        const ageInDays = (now - fileDate) / (1000 * 60 * 60 * 24);
        const ageInYears = ageInDays / 365;

        if (ageInYears >= 5) return `${Math.floor(ageInYears)} years (Legacy)`;
        if (ageInYears >= 3) return `${Math.floor(ageInYears)} years (Old)`;
        if (ageInYears >= 1) return `${Math.floor(ageInYears)} year${Math.floor(ageInYears) > 1 ? 's' : ''} (Mature)`;

        const ageInMonths = ageInDays / 30;
        if (ageInMonths >= 3) return `${Math.floor(ageInMonths)} months (Recent)`;
        return `${Math.floor(ageInDays)} days (New)`;
      })()
    : 'Unknown';

  let detailsHtml = `
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">${file.path}</span>
    </div>
    ${githubFileUrl ? `<div class="info-row">
      <span class="label">View on GitHub</span>
      <span class="value"><a href="${githubFileUrl}" target="_blank" style="color: #4a9eff; text-decoration: none;">ðŸ”— Open file</a></span>
    </div>` : ''}
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">${file.loc.toLocaleString()}</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.${file.extension}</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">${lastModifiedStr}</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">${authorStr}</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">${commitHashStr}</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">${file.commitCount !== null ? `${file.commitCount} commit${file.commitCount !== 1 ? 's' : ''}` : 'Unknown'}</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">${file.contributorCount !== null ? file.contributorCount : 'Unknown'}</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">${fileAgeStr}</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">${file.recentLinesChanged !== null ? `${file.recentLinesChanged} lines changed` : 'Unknown'}</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">${file.avgLinesPerCommit !== null ? `${file.avgLinesPerCommit} lines/commit` : 'Unknown'}</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">${file.daysSinceLastModified !== null ? `${file.daysSinceLastModified} days ago` : 'Unknown'}</span>
    </div>
  `;

  // Add commit info section if available
  if (commitInfo && (commitInfo.message || commitInfo.siblings.length > 0)) {
    detailsHtml += `
          <div style="margin-top: 15px; padding-top: 15px; border-top: 1px solid rgba(255, 255, 255, 0.1);">
        `;

    // Show commit info if message is available
    if (commitInfo.message) {
      detailsHtml += `
            <div style="margin-bottom: 12px;">
              <div style="font-size: 11px; color: #888; margin-bottom: 4px;">
                Commit: <span style="color: #4a9eff; font-family: monospace;">${commitInfo.commitHashStr}</span>
              </div>
              <div style="font-size: 12px; color: #ddd; font-style: italic; line-height: 1.4;">
                "${commitInfo.message}"
              </div>
            </div>
          `;
    }

    // Show commit siblings if any
    if (commitInfo.siblings.length > 0) {
      detailsHtml += `
            <div style="font-size: 12px; font-weight: 600; color: #4a9eff; margin-bottom: 8px;">
              Commit Siblings (${commitInfo.siblings.length} file${commitInfo.siblings.length !== 1 ? 's' : ''})
            </div>
            <div style="font-size: 11px;">
          `;

      for (const sibling of commitInfo.siblings) {
        detailsHtml += `
              <div style="padding: 4px 0; color: #ccc; border-bottom: 1px solid rgba(255, 255, 255, 0.05);">
                ${sibling.path}
              </div>
            `;
      }

      detailsHtml += `
            </div>
          `;
    }

    detailsHtml += `
          </div>
        `;
  }

  // Add coupling analysis section if available
  if (clusterInfo) {
    const { cluster, topEdges } = clusterInfo;

    detailsHtml += `
            <div style="margin-top: 15px; padding-top: 15px; border-top: 1px solid rgba(255, 255, 255, 0.1);">
              <div style="font-size: 12px; font-weight: 600; color: #4a9eff; margin-bottom: 8px;">
                ðŸ“Š Coupling Cluster
              </div>
              <div style="font-size: 11px; color: #ccc; margin-bottom: 12px; padding-left: 12px;">
                ${cluster.name} (${cluster.fileCount} files)
              </div>
          `;

    if (topEdges.length > 0) {
      detailsHtml += `
              <div style="font-size: 12px; font-weight: 600; color: #4a9eff; margin-bottom: 8px;">
                ðŸ”— Most Frequently Changes With
              </div>
              <div style="font-size: 11px;">
            `;

      for (const edge of topEdges) {
        const otherFile = edge.fileA === file.path ? edge.fileB : edge.fileA;
        const fileName = otherFile.split('/').pop() || otherFile;
        const couplingPercent = Math.round(edge.coupling * 100);

        detailsHtml += `
                <div style="padding: 4px 0; color: #ccc; border-bottom: 1px solid rgba(255, 255, 255, 0.05);">
                  <div style="display: flex; justify-content: space-between; align-items: center;">
                    <span style="color: #ddd;">${fileName}</span>
                    <span style="color: #888; font-size: 10px;">${edge.coChangeCount} co-changes</span>
                  </div>
                  <div style="font-size: 10px; color: #888; margin-top: 2px;">
                    ${couplingPercent}% coupling strength
                  </div>
                </div>
              `;
      }

      detailsHtml += `
              </div>
            `;
    }

    detailsHtml += `
            </div>
          `;
  }

  return detailsHtml;
}
