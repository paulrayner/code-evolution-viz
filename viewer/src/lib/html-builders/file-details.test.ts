import { describe, it, expect } from 'vitest';
import { buildFileDetailsHTML, FileDetailsData } from './file-details';
import { FileNode } from '../../types';
import { Cluster, CouplingEdge } from '../../coupling-types';

// Helper function to create mock file nodes
function createMockFile(overrides: Partial<FileNode> = {}): FileNode {
  return {
    path: overrides.path || '/src/utils/test.ts',
    name: overrides.name || 'test.ts',
    type: 'file',
    loc: overrides.loc ?? 250,
    extension: overrides.extension || 'ts',
    lastModified: overrides.lastModified || '2025-10-15T10:30:00Z',
    lastAuthor: overrides.lastAuthor || 'John Doe',
    lastCommitHash: overrides.lastCommitHash || 'abc123def456789',
    commitCount: overrides.commitCount ?? 15,
    contributorCount: overrides.contributorCount ?? 3,
    firstCommitDate: overrides.firstCommitDate || '2023-01-15T08:00:00Z',
    recentLinesChanged: overrides.recentLinesChanged ?? 120,
    avgLinesPerCommit: overrides.avgLinesPerCommit ?? 25,
    daysSinceLastModified: overrides.daysSinceLastModified ?? 8,
  };
}

describe('buildFileDetailsHTML', () => {
  it('should render complete file details with all data present', () => {
    const file = createMockFile();
    const data: FileDetailsData = {
      file,
      githubFileUrl: 'https://github.com/test/repo/blob/main/src/utils/test.ts',
      commitInfo: {
        commitHashStr: 'abc123d',
        message: 'Add new utility function for data transformation',
        siblings: [
          createMockFile({ path: '/src/utils/helper.ts', name: 'helper.ts' }),
          createMockFile({ path: '/src/index.ts', name: 'index.ts' }),
        ],
      },
      clusterInfo: {
        cluster: {
          id: 1,
          name: 'Utilities Cluster',
          files: ['/src/utils/test.ts', '/src/utils/helper.ts'],
          fileCount: 15,
          avgInternalCoupling: 0.75,
        },
        topEdges: [
          {
            fileA: '/src/utils/test.ts',
            fileB: '/src/utils/parser.ts',
            coChangeCount: 42,
            coupling: 0.85,
          },
          {
            fileA: '/src/utils/test.ts',
            fileB: '/src/utils/validator.ts',
            coChangeCount: 28,
            coupling: 0.62,
          },
        ],
      },
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toMatchSnapshot();
  });

  it('should render file details without GitHub URL', () => {
    const file = createMockFile();
    const data: FileDetailsData = {
      file,
      githubFileUrl: null,
      commitInfo: null,
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toMatchSnapshot();
    expect(html).not.toContain('View on GitHub');
  });

  it('should render file details without coupling data', () => {
    const file = createMockFile();
    const data: FileDetailsData = {
      file,
      githubFileUrl: 'https://github.com/test/repo/blob/main/src/utils/test.ts',
      commitInfo: {
        commitHashStr: 'abc123d',
        message: 'Update test file',
        siblings: [],
      },
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toMatchSnapshot();
    expect(html).not.toContain('Coupling Cluster');
    expect(html).not.toContain('Most Frequently Changes With');
  });

  it('should render file details without commit info', () => {
    const file = createMockFile();
    const data: FileDetailsData = {
      file,
      githubFileUrl: 'https://github.com/test/repo/blob/main/src/utils/test.ts',
      commitInfo: null,
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toMatchSnapshot();
    expect(html).not.toContain('Commit Siblings');
  });

  it('should render commit siblings without commit message', () => {
    const file = createMockFile();
    const data: FileDetailsData = {
      file,
      githubFileUrl: null,
      commitInfo: {
        commitHashStr: 'abc123d',
        message: '',
        siblings: [
          createMockFile({ path: '/src/app.ts', name: 'app.ts' }),
        ],
      },
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toMatchSnapshot();
    expect(html).toContain('Commit Siblings');
    expect(html).not.toContain('Commit: <span');
  });

  it('should render commit message without siblings', () => {
    const file = createMockFile();
    const data: FileDetailsData = {
      file,
      githubFileUrl: null,
      commitInfo: {
        commitHashStr: 'def456a',
        message: 'Fix critical bug in parser',
        siblings: [],
      },
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toMatchSnapshot();
    expect(html).toContain('Fix critical bug in parser');
    expect(html).not.toContain('Commit Siblings');
  });

  it('should handle file with minimal metadata (nulls)', () => {
    const file: FileNode = {
      path: '/src/minimal.ts',
      name: 'minimal.ts',
      type: 'file',
      loc: 100,
      extension: 'ts',
      lastModified: null,
      lastAuthor: null,
      lastCommitHash: null,
      commitCount: null,
      contributorCount: null,
      firstCommitDate: null,
      recentLinesChanged: null,
      avgLinesPerCommit: null,
      daysSinceLastModified: null,
    };

    const data: FileDetailsData = {
      file,
      githubFileUrl: null,
      commitInfo: null,
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toMatchSnapshot();
    expect(html).toContain('Unknown');
  });

  it('should format file age correctly for legacy files (5+ years)', () => {
    const fiveYearsAgo = new Date();
    fiveYearsAgo.setFullYear(fiveYearsAgo.getFullYear() - 6);

    const file = createMockFile({
      firstCommitDate: fiveYearsAgo.toISOString(),
    });

    const data: FileDetailsData = {
      file,
      githubFileUrl: null,
      commitInfo: null,
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toContain('years (Legacy)');
  });

  it('should format file age correctly for old files (3-5 years)', () => {
    const fourYearsAgo = new Date();
    fourYearsAgo.setFullYear(fourYearsAgo.getFullYear() - 4);

    const file = createMockFile({
      firstCommitDate: fourYearsAgo.toISOString(),
    });

    const data: FileDetailsData = {
      file,
      githubFileUrl: null,
      commitInfo: null,
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toContain('years (Old)');
  });

  it('should format file age correctly for recent files (months)', () => {
    const fourMonthsAgo = new Date();
    fourMonthsAgo.setMonth(fourMonthsAgo.getMonth() - 4);

    const file = createMockFile({
      firstCommitDate: fourMonthsAgo.toISOString(),
    });

    const data: FileDetailsData = {
      file,
      githubFileUrl: null,
      commitInfo: null,
      clusterInfo: null,
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toContain('months (Recent)');
  });

  it('should render coupling cluster with no edges', () => {
    const file = createMockFile();
    const data: FileDetailsData = {
      file,
      githubFileUrl: null,
      commitInfo: null,
      clusterInfo: {
        cluster: {
          id: 2,
          name: 'Isolated Cluster',
          files: ['/src/utils/test.ts'],
          fileCount: 1,
          avgInternalCoupling: 0.0,
        },
        topEdges: [],
      },
    };

    const html = buildFileDetailsHTML(data);
    expect(html).toMatchSnapshot();
    expect(html).toContain('Coupling Cluster');
    expect(html).toContain('Isolated Cluster');
    expect(html).not.toContain('Most Frequently Changes With');
  });

  it('should render commit count pluralization correctly', () => {
    const fileOne = createMockFile({ commitCount: 1 });
    const fileMany = createMockFile({ commitCount: 42 });

    const dataOne: FileDetailsData = {
      file: fileOne,
      githubFileUrl: null,
      commitInfo: null,
      clusterInfo: null,
    };

    const dataMany: FileDetailsData = {
      file: fileMany,
      githubFileUrl: null,
      commitInfo: null,
      clusterInfo: null,
    };

    const htmlOne = buildFileDetailsHTML(dataOne);
    const htmlMany = buildFileDetailsHTML(dataMany);

    expect(htmlOne).toContain('1 commit<');
    expect(htmlMany).toContain('42 commits<');
  });
});
