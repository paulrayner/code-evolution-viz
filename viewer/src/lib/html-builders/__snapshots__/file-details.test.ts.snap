// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildFileDetailsHTML > should handle file with minimal metadata (nulls) 1`] = `
"
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">/src/minimal.ts</span>
    </div>
    
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">100</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.ts</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">Unknown</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">Unknown</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">Unknown</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">Unknown</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">Unknown</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">Unknown</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">Unknown</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">Unknown</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">Unknown</span>
    </div>
  "
`;

exports[`buildFileDetailsHTML > should render commit message without siblings 1`] = `
"
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">/src/utils/test.ts</span>
    </div>
    
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">250</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.ts</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">Oct 15, 2025</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">John Doe</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">abc123d</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">15 commits</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">3</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">2 years (Mature)</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">120 lines changed</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">25 lines/commit</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">8 days ago</span>
    </div>
  
          <div style="margin-top: 15px; padding-top: 15px; border-top: 1px solid rgba(255, 255, 255, 0.1);">
        
            <div style="margin-bottom: 12px;">
              <div style="font-size: 11px; color: #888; margin-bottom: 4px;">
                Commit: <span style="color: #4a9eff; font-family: monospace;">def456a</span>
              </div>
              <div style="font-size: 12px; color: #ddd; font-style: italic; line-height: 1.4;">
                "Fix critical bug in parser"
              </div>
            </div>
          
          </div>
        "
`;

exports[`buildFileDetailsHTML > should render commit siblings without commit message 1`] = `
"
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">/src/utils/test.ts</span>
    </div>
    
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">250</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.ts</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">Oct 15, 2025</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">John Doe</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">abc123d</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">15 commits</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">3</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">2 years (Mature)</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">120 lines changed</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">25 lines/commit</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">8 days ago</span>
    </div>
  
          <div style="margin-top: 15px; padding-top: 15px; border-top: 1px solid rgba(255, 255, 255, 0.1);">
        
            <div style="font-size: 12px; font-weight: 600; color: #4a9eff; margin-bottom: 8px;">
              Commit Siblings (1 file)
            </div>
            <div style="font-size: 11px;">
          
              <div style="padding: 4px 0; color: #ccc; border-bottom: 1px solid rgba(255, 255, 255, 0.05);">
                /src/app.ts
              </div>
            
            </div>
          
          </div>
        "
`;

exports[`buildFileDetailsHTML > should render complete file details with all data present 1`] = `
"
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">/src/utils/test.ts</span>
    </div>
    <div class="info-row">
      <span class="label">View on GitHub</span>
      <span class="value"><a href="https://github.com/test/repo/blob/main/src/utils/test.ts" target="_blank" style="color: #4a9eff; text-decoration: none;">ðŸ”— Open file</a></span>
    </div>
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">250</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.ts</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">Oct 15, 2025</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">John Doe</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">abc123d</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">15 commits</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">3</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">2 years (Mature)</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">120 lines changed</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">25 lines/commit</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">8 days ago</span>
    </div>
  
          <div style="margin-top: 15px; padding-top: 15px; border-top: 1px solid rgba(255, 255, 255, 0.1);">
        
            <div style="margin-bottom: 12px;">
              <div style="font-size: 11px; color: #888; margin-bottom: 4px;">
                Commit: <span style="color: #4a9eff; font-family: monospace;">abc123d</span>
              </div>
              <div style="font-size: 12px; color: #ddd; font-style: italic; line-height: 1.4;">
                "Add new utility function for data transformation"
              </div>
            </div>
          
            <div style="font-size: 12px; font-weight: 600; color: #4a9eff; margin-bottom: 8px;">
              Commit Siblings (2 files)
            </div>
            <div style="font-size: 11px;">
          
              <div style="padding: 4px 0; color: #ccc; border-bottom: 1px solid rgba(255, 255, 255, 0.05);">
                /src/utils/helper.ts
              </div>
            
              <div style="padding: 4px 0; color: #ccc; border-bottom: 1px solid rgba(255, 255, 255, 0.05);">
                /src/index.ts
              </div>
            
            </div>
          
          </div>
        
            <div style="margin-top: 15px; padding-top: 15px; border-top: 1px solid rgba(255, 255, 255, 0.1);">
              <div style="font-size: 12px; font-weight: 600; color: #4a9eff; margin-bottom: 8px;">
                ðŸ“Š Coupling Cluster
              </div>
              <div style="font-size: 11px; color: #ccc; margin-bottom: 12px; padding-left: 12px;">
                Utilities Cluster (15 files)
              </div>
          
              <div style="font-size: 12px; font-weight: 600; color: #4a9eff; margin-bottom: 8px;">
                ðŸ”— Most Frequently Changes With
              </div>
              <div style="font-size: 11px;">
            
                <div style="padding: 4px 0; color: #ccc; border-bottom: 1px solid rgba(255, 255, 255, 0.05);">
                  <div style="display: flex; justify-content: space-between; align-items: center;">
                    <span style="color: #ddd;">parser.ts</span>
                    <span style="color: #888; font-size: 10px;">42 co-changes</span>
                  </div>
                  <div style="font-size: 10px; color: #888; margin-top: 2px;">
                    85% coupling strength
                  </div>
                </div>
              
                <div style="padding: 4px 0; color: #ccc; border-bottom: 1px solid rgba(255, 255, 255, 0.05);">
                  <div style="display: flex; justify-content: space-between; align-items: center;">
                    <span style="color: #ddd;">validator.ts</span>
                    <span style="color: #888; font-size: 10px;">28 co-changes</span>
                  </div>
                  <div style="font-size: 10px; color: #888; margin-top: 2px;">
                    62% coupling strength
                  </div>
                </div>
              
              </div>
            
            </div>
          "
`;

exports[`buildFileDetailsHTML > should render coupling cluster with no edges 1`] = `
"
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">/src/utils/test.ts</span>
    </div>
    
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">250</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.ts</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">Oct 15, 2025</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">John Doe</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">abc123d</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">15 commits</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">3</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">2 years (Mature)</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">120 lines changed</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">25 lines/commit</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">8 days ago</span>
    </div>
  
            <div style="margin-top: 15px; padding-top: 15px; border-top: 1px solid rgba(255, 255, 255, 0.1);">
              <div style="font-size: 12px; font-weight: 600; color: #4a9eff; margin-bottom: 8px;">
                ðŸ“Š Coupling Cluster
              </div>
              <div style="font-size: 11px; color: #ccc; margin-bottom: 12px; padding-left: 12px;">
                Isolated Cluster (1 files)
              </div>
          
            </div>
          "
`;

exports[`buildFileDetailsHTML > should render file details without GitHub URL 1`] = `
"
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">/src/utils/test.ts</span>
    </div>
    
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">250</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.ts</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">Oct 15, 2025</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">John Doe</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">abc123d</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">15 commits</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">3</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">2 years (Mature)</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">120 lines changed</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">25 lines/commit</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">8 days ago</span>
    </div>
  "
`;

exports[`buildFileDetailsHTML > should render file details without commit info 1`] = `
"
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">/src/utils/test.ts</span>
    </div>
    <div class="info-row">
      <span class="label">View on GitHub</span>
      <span class="value"><a href="https://github.com/test/repo/blob/main/src/utils/test.ts" target="_blank" style="color: #4a9eff; text-decoration: none;">ðŸ”— Open file</a></span>
    </div>
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">250</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.ts</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">Oct 15, 2025</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">John Doe</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">abc123d</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">15 commits</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">3</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">2 years (Mature)</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">120 lines changed</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">25 lines/commit</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">8 days ago</span>
    </div>
  "
`;

exports[`buildFileDetailsHTML > should render file details without coupling data 1`] = `
"
    <div class="info-row">
      <span class="label">Type</span>
      <span class="value">File</span>
    </div>
    <div class="info-row">
      <span class="label">Path</span>
      <span class="value">/src/utils/test.ts</span>
    </div>
    <div class="info-row">
      <span class="label">View on GitHub</span>
      <span class="value"><a href="https://github.com/test/repo/blob/main/src/utils/test.ts" target="_blank" style="color: #4a9eff; text-decoration: none;">ðŸ”— Open file</a></span>
    </div>
    <div class="info-row">
      <span class="label">Lines of Code</span>
      <span class="value">250</span>
    </div>
    <div class="info-row">
      <span class="label">Extension</span>
      <span class="value">.ts</span>
    </div>
    <div class="info-row">
      <span class="label">Last Modified</span>
      <span class="value">Oct 15, 2025</span>
    </div>
    <div class="info-row">
      <span class="label">Last Author</span>
      <span class="value">John Doe</span>
    </div>
    <div class="info-row">
      <span class="label">Last Commit</span>
      <span class="value">abc123d</span>
    </div>
    <div class="info-row">
      <span class="label">Churn (Lifetime)</span>
      <span class="value">15 commits</span>
    </div>
    <div class="info-row">
      <span class="label">Contributors (Lifetime)</span>
      <span class="value">3</span>
    </div>
    <div class="info-row">
      <span class="label">File Age</span>
      <span class="value">2 years (Mature)</span>
    </div>
    <div class="info-row">
      <span class="label">Recent Activity (90 days)</span>
      <span class="value">120 lines changed</span>
    </div>
    <div class="info-row">
      <span class="label">Avg Change Size (Lifetime)</span>
      <span class="value">25 lines/commit</span>
    </div>
    <div class="info-row">
      <span class="label">Last Touched</span>
      <span class="value">8 days ago</span>
    </div>
  
          <div style="margin-top: 15px; padding-top: 15px; border-top: 1px solid rgba(255, 255, 255, 0.1);">
        
            <div style="margin-bottom: 12px;">
              <div style="font-size: 11px; color: #888; margin-bottom: 4px;">
                Commit: <span style="color: #4a9eff; font-family: monospace;">abc123d</span>
              </div>
              <div style="font-size: 12px; color: #ddd; font-style: italic; line-height: 1.4;">
                "Update test file"
              </div>
            </div>
          
          </div>
        "
`;
